overlord/snapstate/snapstate.go

```diff
@@ -451,7 +451,9 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 				retain = 3
 			}
 		}
-		retain-- //  we're adding one
+		if snapst.LastIndex(snapsup.Revision()) == -1 {
+			retain-- //  we're adding one
+		}
```

Comment:
Could you add a comment here explaining why we need this extra check (basically, what you explained in the PR description could be reused here).

Reply:
snapsup.Revision() is copied to  targetRevision earlier, also there's a loop below that seems to be checking a similar thing to the new "if" here, so perhaps this should be integrated there somehow?

```
		// make sure we're not scheduling the removal of the target
		// revision in the case where the target revision is already in
		// the sequence.
		for i := 0; i < currentIndex; i++ {
```